pkg/target/manager.go

```diff
@@ -282,8 +292,122 @@ func (m *managerImpl) validateSeed(ctx context.Context, seed *gardencorev1beta1.
 	return nil
 }
 
+func domainsFromConfiguration(m *managerImpl) (map[string]string, map[string]string, error) {
+	domainRegexp, err := regexp.Compile(`(http://|https://)?(.+\..+)`)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to compile regular expression: %v", err)
+	}
+
+	dashboardDomains := make(map[string]string)
+	shootDomains := make(map[string]string)
+
+	for _, g := range m.Configuration().Gardens {
+		match := domainRegexp.FindStringSubmatch(g.DashboardDomain)
+		if match != nil {
+			dashboardDomains[g.Name] = match[2]
+		}
+
+		match = domainRegexp.FindStringSubmatch(g.ShootDomain)
+		if match != nil {
+			shootDomains[g.Name] = match[2]
+		}
+	}
+
+	return dashboardDomains, shootDomains, nil
+}
+
+func matchGardenDomain(domains map[string]string, domain string) (string, string) {
+	for garden, gDomain := range domains {
+		if strings.Contains(domain, gDomain) {
+			return garden, domain
+		}
+	}
+
+	return "", ""
+}
+
+func targetFlagsFromDomain(ctx context.Context, m *managerImpl, t *targetImpl, name string) (TargetFlags, error) {
+	dashboardDomains, shootDomains, err := domainsFromConfiguration(m)
+	if err != nil {
+		return nil, fmt.Errorf("invalid domain format in configured gardens: %w", err)
+	}
+
+	if len(dashboardDomains) == 0 && len(shootDomains) == 0 {
+		return nil, fmt.Errorf("no domain found in configured gardens. Shoot targeting via domain requires shoot and / or dashboard domain configuration for each garden that you want to enable domain targeting for")
+	}
+
+	garden, domain := matchGardenDomain(dashboardDomains, name)
+	if domain != "" {
+		domainRegexp, err := regexp.Compile(`.+/namespace/([^/]+)/shoots/([^/]+)/?`)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract project and shoot from url: %w", err)
+		}
+
+		match := domainRegexp.FindStringSubmatch(domain)
+		if match != nil {
+			gardenClient, err := m.clientForGarden(t.Garden)
+			if err != nil {
+				return nil, fmt.Errorf("could not create Kubernetes client for garden cluster: %w", err)
+			}
+
+			namespace, err := m.resolveProjectNamespace(ctx, gardenClient, match[1])
+			if err != nil {
+				return nil, fmt.Errorf("failed to validate project: %w", err)
+			}
+
+			if namespace == nil {
+				return nil, fmt.Errorf("invalid namespace in shoot url")
+			}
+
+			projectName := namespace.Labels["project.gardener.cloud/name"]
+			if projectName == "" {
+				return nil, fmt.Errorf("namespace in url is not related to a gardener project")
+			}
+
+			return NewTargetFlags(garden, projectName, "", match[2]), nil
+		}
+	}
+
+	garden, domain = matchGardenDomain(shootDomains, name)
+	if domain != "" {
+		domainRegexp, err := regexp.Compile(`([^.]+).([^.]+)\..+`)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract project and shoot from url: %w", err)
+		}
+
+		match := domainRegexp.FindStringSubmatch(domain)
+		if match != nil {
+			return NewTargetFlags(garden, match[2], "", match[1]), nil
```

Comment:
I think it makes it hard to read what match[2] and match[1] really means, it probably makes more sense to extract it to a variable and name it appropriately